This notebook containts the following:

01) Finding and dealing with missing values

02) Finding and dealing with duplicates



01. Setting up notebook

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

02. Establishing a path to import data sets

In [2]:
path = r'C:\Users\Paola\Desktop\Shaul\Data\CF\August-2023 Instacart Basket Analysis'

03. Importing data sets

In [3]:
# Importing 'products.csv' data set
df_prods = pd.read_csv(os.path.join(path, '2-Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Importing 'orders_wrangled.csv' data set
df_ords = pd.read_csv(os.path.join(path, '2-Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

04. Starting Task 4.5 exercises

In [5]:
# Repeating df_ords.describe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

05. Starting dataframe for test/practice

In [6]:
# Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix']=['a', 'b', 1, True]

In [8]:
# Checking 'mix' column contents
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Changing 'mix' column data type => to string (text)
df_test['mix'] = df_test['mix'].astype('str')

In [ ]:
# Changing 'mix' column data type => to integers (numbers)
df_test['mix'] = df_test['mix'].astype('int64')

06. Working on missing values

In [34]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [35]:
# Creating a dataframe to find out the NaN entries
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [36]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

07. Removing missing values

In [49]:
df_prods.shape

(49693, 5)

In [50]:
# Creating a new dataframe without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [51]:
# The rows with missing values have been removed.
df_prods_clean.shape

(49677, 5)

08. Duplicates

In [52]:
# Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [54]:
# Checking shape of df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [55]:
# Removing the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [56]:
df_prods_clean_no_dups.shape

(49672, 5)

09. Exporting dataframe into computer

In [57]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '2-Data','Prepared Data', 'products_checked.csv'))

10. Starting Task 4.5 questions

In [58]:
# Checking df_ords for consistency
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [60]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# Unnamed column is actually the index column, created automatically by Python. 
For statistical purposes, I would ignore the columns 'order_id' and 'order_number'. Although they have 'integer' as data type, their numbers are for reference only and have no statistical value. The column 'used_id' could be grouped by users and compare the difference in behaviour between those customers. However, as far as data consistency is concerned, I cannot pinpoint anything wrong. The columns 'order day of the week' and 'order hour of the day' appear to be ok. I would investigate further the column ' days since prior order': if the values are 0, 4, 7, 15 and 30 days, it seems there is something wrong with the 'max' value. 75% of the data in that column is between '0' and '15', so the data is skewed to the right.


In [62]:
# Checking df_ords for mixing data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No mixed data found in df_ords dataframe

In [63]:
# Looking for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the 'days since prior order' columns. This makes sense, following my abovementioned comments, that the 'max' value seemed a bit strange.

In [64]:
# Viewing the missing values
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [65]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

The missing data is only 0.06% of the total data. Upon observation, it seems it affects the time-period between 11AM-15PM which is the busiest time of the day. However, since this proportion will not have an immediate effect in the overal analysis, I will use imputation to address this issue.

In [66]:
# Will try imputing the missing values. Looking for the median, to decide between it and the mean.
df_ords.median()

C:\Users\Paola\AppData\Local\Temp\ipykernel_8656\4289067234.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_ords.median()


Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

The mean and median have the same value in this case, which is '7'.

Tutor: I have decided to use imputation since the proportion of missing values is less than 0,1%. However, when i used the command, I forgot to adjust the mean value so i actually created a mixed data column... i have left it as is for you to see. As you can se below, I have looked online and used the .replace() function to replace 'mean value' with the actual value which is '7'. It costed me some time, but was worth it as i learned something new.

In [73]:
# Imputing the missing values using the mean.
df_ords['days_since_prior_order'].fillna('mean value', inplace=True)

In [74]:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

Missing values issue resolved.

In [75]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  
count        3.421083e+06       3.421083e+06  
mean         2.776219e+00       1.345202e+01  
std          2.046829e+00       4.226088e+00  
min          0.000000e+00       0.000000e+00  
25%          1.000000e+00       1.000000e+01  
50%          3.000000e+00       1.300000e+01  
75%          5.000000e+00       1.600000e+01  
max          6.000000e+00       2.300000e+01

It is possible to see that values such as the standard deviation, mean and max have been updated after the imputing process.

In [76]:
df_ords.median()

C:\Users\Paola\AppData\Local\Temp\ipykernel_8656\604826045.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_ords.median()


Unnamed: 0            1710541.0
order_id              1710542.0
user_id                102689.0
order_number               11.0
orders_day_of_week          3.0
order_hour_of_day          13.0
dtype: float64

I realised at this point what my error was. i noticed that the 'day since prior order' was missing from the output.

In [77]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

days_since_prior_order


In [79]:
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].replace('mean value', 7)

In [80]:
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
eval_set                  0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [81]:
# Looking for mixed data. Issue solved.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [82]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.421083e+06  
mean         2.776219e+00       1.345202e+01            1.086681e+01  
std          2.046829e+00       4.226088e+00            8.978521e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            5.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

It is possible to see that the 25% and the 75% quartiles, as well as the standard deviation, have been updated.

In [84]:
# Looking for duplicates
df_dups = df_ords[df_ords.duplicated()]

No duplicates found in df_ords

Exporting dataframe (df_ords) into computer

In [85]:
# Exporting 'df_ords' to csv
df_ords.to_csv(os.path.join(path, '2-Data','Prepared Data', 'orders_checked.csv'))
